File: frontend/types/afeSnapshots.ts
export interface AfeSnapshotAttempt {
  id: string
  estimate_version_id: string
  resulting_snapshot_id: string | null
  requested_reference: string | null
  status: 'completed' | 'blocked' | 'denied' | 'failed'
  message: string | null
  eligibility_snapshot: Record<string, unknown>
  created_at: string
  created_by: string | null
}

export interface AfeSnapshotLine {
  id: string
  line_number: number
  item_code: string
  item_description: string
  cost_code: string
  quantity: string
  unit_code: string
  total_cost: string
}

export interface AfeSnapshot {
  id: string
  afe_number: string
  snapshot_type: 'baseline'
  estimate_version_id: string
  calculation_run_id: string
  issue_date: string
  currency_code: string
  grand_total: string
  engine_version: string
  rule_set_version: string
  lines: AfeSnapshotLine[]
  created_at: string
  created_by: string | null
}

export interface EstimateAfeStatus {
  estimate_id: string
  estimate_version_id: string
  version_number: number
  afe_status: 'policy_pending' | 'issued'
  baseline_snapshot: AfeSnapshot | null
  creation_attempts: AfeSnapshotAttempt[]
  pending_requirements: string[]
}

File: backend/app/schemas/afe_snapshots.py
"""Phase 7 immutable baseline AFE snapshot API contracts."""

from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class AfeSnapshotCreateRequest(BaseModel):
    version_id: UUID | None = None
    requested_reference: str | None = Field(default=None, max_length=100)


class AfeSnapshotAttemptRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    estimate_version_id: UUID
    resulting_snapshot_id: UUID | None
    requested_reference: str | None
    status: str
    message: str | None
    eligibility_snapshot: dict[str, Any]
    created_at: datetime
    updated_at: datetime
    created_by: UUID | None
    updated_by: UUID | None


class AfeSnapshotLineRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source_estimate_item_id: UUID
    line_number: int
    item_code: str
    item_description: str
    item_type: str
    cost_code: str
    cost_category_code: str | None
    vendor_code: str | None
    quantity: Decimal
    unit_code: str
    rate_amount: Decimal | None
    rate_currency_code: str | None
    base_cost: Decimal
    contingency_cost: Decimal
    escalation_cost: Decimal
    total_cost: Decimal


class AfeSnapshotRead(BaseModel):
    id: UUID
    afe_number: str
    snapshot_type: str
    estimate_version_id: UUID
    calculation_run_id: UUID
    issue_date: date
    estimate_code: str
    estimate_title: str
    afe_code: str
    project_code: str
    well_code: str
    currency_code: str
    engine_version: str
    rule_set_version: str
    base_total: Decimal
    contingency_total: Decimal
    escalation_total: Decimal
    grand_total: Decimal
    source_snapshot: dict[str, Any]
    lines: list[AfeSnapshotLineRead]
    created_at: datetime
    updated_at: datetime
    created_by: UUID | None
    updated_by: UUID | None


class EstimateAfeStatus(BaseModel):
    estimate_id: UUID
    estimate_version_id: UUID
    version_number: int
    afe_status: str
    baseline_snapshot: AfeSnapshotRead | None
    creation_attempts: list[AfeSnapshotAttemptRead]
    pending_requirements: list[str]
